### DISCUSSION TITLE
onRecordBeforeAuthWithOAuth2Request hook won't trigger

### DISCUSSION THREAD

**Discussion by dimitarnikolovv** [pocketbase/pocketbase#3854] [Q&A]
I am making an `authWithOAuth2Code` request on my server with the intention of setting additional data through the `onRecordBeforeAuthWithOAuth2Request` hook, but for some reason, the hook will not be triggered after the request is made.
And I am getting this error: 

![Screenshot 2023-12-02 at 4 05 06](https://github.com/pocketbase/pocketbase/assets/49731322/ae84c488-f1f7-40d9-b512-2586cecaa311)

The API point where the request is made:

```js
export const GET = async ({ locals, url, cookies }) => {
  const redirectURL = `${url.origin}/api/oauth2`;
  const expectedState = cookies.get('state');
  const expectedVerifier = cookies.get('verifier');

  const state = await url.searchParams.get('state');
  const code = await url.searchParams.get('code');

  const authMethods = await locals.pb.collection('users').listAuthMethods();

  const provider = authMethods.authProviders[0];

  if (expectedState !== state) {
    throw redirect(303, '/login');
  }

  try {
    await locals.pb
      .collection('users')
      .authWithOAuth2Code(provider.name, code, expectedVerifier, redirectURL);
  } catch (err) {
    console.log(err);
  }

  throw redirect(303, '/');
};
```

handle_google_login.pb.js
```js
onRecordBeforeAuthWithOAuth2Request((e) => {
  console.log(JSON.stringify(e.record));
  console.log(JSON.stringify(e.oAuth2User));
  console.log('Is New: ', JSON.stringify(e.isNewRecord));

  const generateUsername = (email) => {
    let id = (Math.random() * 10 + Math.random() * 10).toString(16);
    id = id.split('.').join('');
    const user = email
      .split('@')[0]
      .split(/[^a-zA-Z0-9]/g)
      .join('');

    return `${user}${id}`;
  };

  if (e.record === null) {
    e.record.setEmail(e.oAuth2User.email);
    e.record.set('firstName', e.oAuth2User.rawUser.given_name);
    e.record.set('lastName', e.oAuth2User.rawUser.family_name);
    e.record.set('gender', 'other');
    e.record.set('age', 0);
    e.record.setVerified(true);
    e.record.setUsername(generateUsername(e.oAuth2User.email));

    $app.dao().saveRecord(e.record);
  }
}, 'users');
```

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the above and the `if (e.record === null)` block.

Note that the before hook fires **before** the provider linking, meaning that `e.record` may not be created yet. If you want to set/overwrite some record fields based on the OAuth2 response, it would be better to attach to the [`onRecordAfterAuthWithOAuth2Request` hook](https://pocketbase.io/docs/js-event-hooks/#onrecordafterauthwithoauth2request) and update its `e.record` (_the after hook always have `e.record`, which is either an existing/previously linked auth record or a newly created one_).

**Reply by dimitarnikolovv** [reply]
I am trying to avoid passing `createData` in `authWithOAuth2Code`. And I wasn't sure if this wont break the request. That's not the main problem, however. The main problem is that neither of the hooks triggers.

**Reply by ganigeorgiev** [reply]
I'm not sure how to reproduce it. What version of PocketBase do you use?
Try to remove any custom hooks code that may interrupt the auth flow and check the console output for any additional error messages.

**Reply by dimitarnikolovv** [reply]
I'm running 0.19.4 and I don't have any other hooks that may mess around with the auth collection or that may interrupt the auth flow. I only know that this error is thrown after this query ``SELECT `users`.* FROM `users` WHERE `users`.`id`='112232966209118299212' LIMIT 1``, where the `id` is the id of google's oAuth2User and I know in fact that this user does not exist in the users collection.

**Reply by ganigeorgiev** [reply]
I've tested it again locally with both newly authenticated user and existing one, and in both cases the hooks fire for me.

I don't know what triggers the above query as we don't make requests against the `users` table with the OAuth2 user's id.

There is something else going on in your environment. Double check your custom hooks code.

If you think that it is an issue with PocketBase, please create a **minimal** reproducible repo or at least provide some steps to reproduce the issue.

**Reply by dimitarnikolovv** [reply]
Okay, so I tried to create a reproducible repo, without the pb_data and when I tested it there the hook worked and as you said it did not query the users table with the OAuth2 id. Everything is still **exactly** the same the only difference between the two instances of Pocketbase is the pb_data folder, which is strange.

**Reply by ganigeorgiev** [reply]
This is really confusing and I'm not sure how to help you debug it without details what custom code you are running.

If you need further help, please share what you are running, otherwise I think we are going in circles.

**Reply by dimitarnikolovv** [reply]
I will create a reproducible repo anyway for you to check as I may be missing something. If necessary I can provide some of the pb_data too.

**Reply by dimitarnikolovv** [reply]
Here is the [repo](https://github.com/dimitarnikolovv/oauth2-pocketbase-strange-behaviour)

**Reply by ganigeorgiev** [reply]
I briefly checked it and I can't think of a reason why the reported query in https://github.com/pocketbase/pocketbase/discussions/3854#discussioncomment-7736090 will get executed.

My guess is that it was a result from either an old code that you didn't show or some partially inserted data from older implementation attempts.

But in all cases my recommendation from earlier about using the [`onRecordAfterAuthWithOAuth2Request`](https://pocketbase.io/docs/js-event-hooks/#onrecordafterauthwithoauth2request) hook remains.

Again, if you think that it is a PocketBase issue, feel free to provide a **minimal** reproducible repo or steps that triggers the above and I'll have a look.

**Reply by SkYNewZ** [reply]
Hello. I think i have the exact same issue. Hooks (before or after) are triggered but even by modifying the `e.Record`, the result in database left unchanged. For testing purpose I make a very basic Pocketbase app.

- I have nothing other than a basic Pocketbase instance
- I have added the `foo (string)` property in my database
- I have tested using a pointer string instead of a string

```go
package main

import (
	"log"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

func main() {
	app := pocketbase.New()

	app.OnRecordAfterAuthWithOAuth2Request().Add(func(e *core.RecordAuthWithOAuth2Event) error {
		log.Println("AFTER")
		e.Record.Set("foo", "bar")
		return nil
	})

	if err := app.Start(); err != nil {
		panic(err)
	}
}

```

<img width="1800" alt="Capture d’écran 2023-12-14 à 22 28 12" src="https://github.com/pocketbase/pocketbase/assets/26750012/5001cace-9d4b-48a3-9e65-45d5dce0e272">

**Reply by ganigeorgiev** [reply]
@SkYNewZ _I don't see how your problem is related to the above discussion. I've replied in your existing one https://github.com/pocketbase/pocketbase/discussions/3945._

**Reply by SkYNewZ** [reply]
> but for some reason, the hook will not be triggered after the request is made.

Sorry for disturbing I though it was related.
